### PR TITLE
Line chart refinements and move to components package, portal tooltip fix

### PR DESCRIPTION
**Pull request by sjcobb**
Overview

- move line chart to ui/components so it can be imported from @perses-dev/components and embedded independently of entire panel
- tooltips rebuilt to use Portals
- bugfix: tooltips no longer linger after moving cursor out of canvas
- rename echarts wrapper component to ECharts.tsx
- add subset of supported events to ECharts.tsx
- refactor line chart to use ECharts.tsx
- refactor line chart from 'time' xAxis to 'category' for more control over symbols (which improves performance), axis formatting, series customization
- new hover state that shows focused series symbols on lines as cursor moves
- yAxis formatting to abbreviate large numbers as T, B, M, or k
- dataZoom and toolbox icons enabled
- dataZoom event handler in LineChart.tsx to emit start and end data (will be used to update URL)
- visual refinements to axis colors, fonts, tooltips, grid padding
- line chart loading state using Skeleton

### CODE REVIEW

File: ui/panels-plugin/src/plugins/line-chart/LineChartContainer.tsx [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+export interface LineChartContainerProps {
+  width: number;
+  height: number;
+}
+
+/**
+ * Draws a LineChart with Apache ECharts for the current running time series.
+ */
+export function LineChartContainer(props: LineChartContainerProps) {
```
Comment (sjcobb) [inline ui/panels-plugin/src/plugins/line-chart/LineChartContainer.tsx:37]: Thoughts on a better name for this? I tried combining this component with LineChartPanel.tsx, but keeping them separate was easier (open to any ideas here)

File: ui/components/src/ECharts.tsx [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import React, { useState, useLayoutEffect } from 'react';
+import debounce from 'lodash/debounce';
+import { ECharts, EChartsCoreOption, init } from 'echarts/core';
+import { Box, SxProps, Theme } from '@mui/material';
+
```
Comment (sjcobb) [inline ui/components/src/ECharts.tsx:18]: ECharts does not export many of the internal types that we need, tracking [this issue](https://github.com/apache/echarts/issues/15284) and will hopefully be able to import most of these eventually

**Comment by Nexucis** [comment]
Looks nice :)!

It would be cool at some point to have a sort of documentation explaining how to use it / integrate it !

**Comment by sjcobb** [comment]
> It would be cool at some point to have a sort of documentation explaining how to use it / integrate it !

Yeah definitely--we've talked about potentially using Storybook for this, but I can make some .md notes in the meantime.

**Review by LukeTillman — APPROVED** [review]
Great job. I left some comments/suggestions inline. Let me know if I can help.

**Comment by Nexucis** [comment]
> Yeah definitely--we've talked about potentially using Storybook for this, but I can make some .md notes in the meantime.

I don't know what is this format, but the usual way in opensource code is to have the doc in markdown and then later you build a website to display this docs so it's nicer to read it. (As markdown can be a bit harsh to read)

File: ui/components/src/tooltip/Tooltip.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+export function Tooltip(props: TooltipProps) {
+  const { chartRef, chartData } = props;
+  const [pinnedPos, setPinnedPos] = useState<CursorCoordinates | null>(null);
+  const mousePos = useMousePosition();
+  if (mousePos === null) return null;
+
+  const chart = chartRef.current;
+  const focusedSeries = getFocusedSeriesData(mousePos, chartData, pinnedPos, chart);
+  const chartWidth = chart?.getWidth() ?? 1000;
+  const chartHeight = chart?.getHeight() ?? 250;
```
[the remark this answers is not in the document]
Comment (sjcobb) [reply ui/components/src/tooltip/Tooltip.tsx:45]: No good reason actually :smile: they were roughly the dimensions of the chart I was replacing. Just updated the defaults to be 750x230, which makes more sense for Perses (where charts are usually half width).

File: ui/panels-plugin/src/plugins/line-chart/LineChartContainer.tsx (hunk 1 above)
Comment (sjcobb) [reply ui/panels-plugin/src/plugins/line-chart/LineChartContainer.tsx:37]: (discussed off-line with Luke) Reason is GraphQueryRunner passes data via context, but the chart needs it as a prop. Short term we're going to leave LineChartContainer as is, until graph queries approach is potentially updated to take advantage of react-query's `useQueries` hook.

File: ui/components/src/tooltip/TooltipContent.tsx [hunk 4]
```diff
@@ -39,17 +44,22 @@ function TooltipContent(props: TooltipContentProps) {
     );
   };
 
-  if (focusedSeries && seriesTime) {
+  if (focusedSeries !== null && seriesTime !== null) {
+    const sortedFocusedSeries = focusedSeries.sort((a, b) => (a.y > b.y ? -1 : 1));
```
[the remark this answers is not in the document]
Comment (sjcobb) [reply ui/components/src/tooltip/TooltipContent.tsx:48]: Yeah good call, just fixed this

File: ui/components/src/ECharts.tsx [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    }, 200);
+    let observer: ResizeObserver | undefined;
+    if (parentContainer != null) {
+      observer = new ResizeObserver(() => {
+        updateSize();
+      });
+      observer.observe(parentContainer);
+    } else {
+      window.addEventListener('resize', updateSize);
+    }
```
[the remark this answers is not in the document]
Comment (sjcobb) [reply ui/components/src/ECharts.tsx:169]: As discussed off-line, this has been reverted to how it was before

**Comment by sjcobb** [comment]
> I don't know what is this format, but the usual way in opensource code is to have the doc in markdown and then later you build a website to display this docs so it's nicer to read it. (As markdown can be a bit harsh to read)

This sounds good @Nexucis, and Storybook would be an enhancement to the docs website--a lot of projects use it for exploring component libraries. For example, GitLab links to [their Storybook](https://gitlab-org.gitlab.io/gitlab/storybook/?path=/docs/usage-quotas-components-statistics-card--default) from [this page](https://docs.gitlab.com/ee/development/fe_guide/storybook.html) on their website. And [here's another example](https://apache-superset.github.io/superset-ui/?path=/story/core-packages-superset-ui-color--categorical-palettes) with more components built out.

**Comment by Nexucis** [comment]
@sjcobb ooooh that's interesting ! Thanks for the explanation, it looks really cool indeed !

**Pull request merged**